### PR TITLE
feat(helm): Adding KEDA autoscaling support

### LINKED ISSUE #3430: Add support for HPA in mimir-distributed
#### Is your feature request related to a problem? Please describe.

I have been running Grafana Mimir in our infrastructure using the Helm chart and I'm not currently able to automatically scale mimir in response to changes in load on the servers.

#### Describe the solution you'd like

It would be very helpful if Horizontal Pod Autoscalers were supported natively in the mimir-distributed helm chart.

#### Describe alternatives you've considered

I've attempted to add this manually using custom manifests with the chart to create HPAs for each component (Distributor, Querier, etc.), however, this is not an optimal solution since the `replicas` field on each Deployment and StatefulSet cannot be removed with the current templating. When using a custom HPA with the helm chart, applying any changes to one of the Deployments/StatefulSets causes a scale down to the number specified by the `replicas` field, followed by a scale up once the HPA has updated the desired replicas. In production, this is not ideal since it will potentially terminate active pods and disrupt our cluster's monitoring. 

As of now, my best solution is to fork the chart and add templating to omit this field, but it would be nice if this could be officially supported.

### PR DESCRIPTION
**Pull request by beatkind**
#### What this PR does

This PR takes #4687 as base layer and tries to implement all added code comments. Please be aware that this PR does not cover the changes in the monitoring dashboards that may be needed.

#### Which issue(s) this PR fixes or relates to

Fixes #3430

#### Checklist

- [x] Tests updated.
- [ ] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

**Comment by beatkind** [comment]
@dimitarvdimitrov I have opened this PR and took #4687 as a base. I have ported the changes from your PR #6971 into here, as you have mentioned in [this](https://github.com/grafana/mimir/pull/4687#issuecomment-1828412906) comment.

I know that there is still documentation missing for this PR. Could you, nevertheless, have a look over the changes?

**Review by ChrisNoSim — APPROVED** [review]

**Comment by dimitarvdimitrov** [comment]
many thanks for picking this up. I will set some time aside this week to review it

**Comment by dimitarvdimitrov** [comment]
Have you had the chance to test this out on a mimir cell? It would be nice to do that before we merge the PR

**Comment by beatkind** [comment]
I did tried it in a minikube setup, but not yet in a real world scenario. 

Do you feel confident in merging it as experimental feature? This would give me the option to test it in our production cluster.

**Review by dimitarvdimitrov — COMMENTED** [review]
First of all, great work @beatkind! Thank you for picking this up. 

### Review


The PR is in a really good shape. I left a few minor comments and I think they'll be quick to address. Can you take a look? I'm happy to merge this as an experimental feature after addressing them.

I apologise for the delayed review, but I couldn't find a large-enough block of time to review this (it's a hefty PR). I'll try to be more responsive for the next iterations.

### Docs

Perhaps the major reason for writing docs is to cover the migration procedure for enabling autoscaling. At Grafana Labs we had to go through a migration procedure to make sure that there aren't any disruptions. The main problem was that removing the `replicas` field causes the deployment to scale to 1 replica if the HPA hasn't had the time to take control, which is very likely if you're removing replicas and creating the HPA in one go.

I think a reasonable migration procedure would be to add a config flag to kedaAutoscaling which preserves `replicas` (e.g. `*.kedaAutoscaling.keepReplicas`). If you'd like to add that in this PR you're welcome, but this PR is already quite large and is already a beneficial addition. I opened this issue to track the migration docs and configuration https://github.com/grafana/mimir/issues/7367

Do you agree with this? Maybe I'm missing something about the migration.

### Future work

There some bits which are missing from this PR which I think we should add before calling this work complete. I've added them in another issue https://github.com/grafana/mimir/issues/7368. As the one opening this PR, would you like to take ownership of these follow-up actions? It's ok to say no, just decided to offer you the chance.


also cc @krajorama in case you want to review this too

**Comment by beatkind** [comment]
@dimitarvdimitrov many thanks for the review! I am happy to add missing parts that you have mentioned.

I also totally agree with the migration path. You are right, this is already a big PR. I will add the things you have mentioned, let then merge it as experimental. I will also add a comment regarding to be careful regarding the mentioned replica situation. 

I am happy to take ownership of the issues you have created and I plan to open another PR once that is merged introducing the migration to autoscaler and adding docs :) 

Thanks you very much!

File: operations/helm/charts/mimir-distributed/templates/query-frontend/query-frontend-so.yaml
Comment (jhesketh) [reply operations/helm/charts/mimir-distributed/templates/query-frontend/query-frontend-so.yaml:27]: > Jsonnet still supplies the metricName field. In jsonnet the _default suffix is actually the namespace. I wonder if we should spare ourselves the trouble and just remove it now from the helm templates.

I'd be for removing it since this is a new feature to helm.

> @jhesketh what's your take on how many keda versions we should support? KEDA 2.10 is from March 2023, 2.12 is from September 2023. How painful are KEDA upgrades? Can we require a fairly up-to-date version of KEDA (11 months old; is that even up to date)?

I don't have a particularly strong opinion here, but since this is a new (to helm) feature, I would tend towards just supporting the latest version of KEDA. The jsonnet may work with older versions, but we should prefer the newest.

I don't believe upgrading KEDA is painful, but I have not done so.

Comment (dimitarvdimitrov) [reply operations/helm/charts/mimir-distributed/templates/query-frontend/query-frontend-so.yaml:27]: ok, then let's drop the metricName from all the ScaledObjects. @beatkind can you do that?

Comment (beatkind) [reply operations/helm/charts/mimir-distributed/templates/query-frontend/query-frontend-so.yaml:27]: I agree here. Only supporting the newest version of KEDA makes sense. I will drop the `metricName`field.

File: operations/helm/charts/mimir-distributed/values.yaml
Comment (beatkind) [reply operations/helm/charts/mimir-distributed/values.yaml:1340]: @dimitarvdimitrov I think you mean 75 and 100?

Comment (beatkind) [reply operations/helm/charts/mimir-distributed/values.yaml:1340]: @dimitarvdimitrov I think you mean 75 and 100?

File: operations/helm/charts/mimir-distributed/templates/metamonitoring/_helpers.tpl
Comment (beatkind) [reply operations/helm/charts/mimir-distributed/templates/metamonitoring/_helpers.tpl:67]: @dimitarvdimitrov how do you feel about introducing a global field to the KEDA remote read URL? In some scenarios, people might leverage a custom deployed Prometheus (or Prometheus operator) for monitoring their mimir setup.

I am thinking about something like `kedaRemoteReadURL`, if unset we are using the metamonitoring URL (`mimir.remoteReadUrl.inCluster`), if set the one provided through the user.

Comment (dimitarvdimitrov) [reply operations/helm/charts/mimir-distributed/templates/metamonitoring/_helpers.tpl:67]: I think that makes sense. It's also some part of the follow-up work issue https://github.com/grafana/mimir/issues/7368. But my comment here was about not parsing the URL of metamonitoring and instead use it verbatim (`{{- (.remote).url -}}`), which still applies even if we add `kedaRemoteReadURL`


#### regarding `kedaRemoteReadURL`

Where would that go? With the current model it would fit under each component

```yaml
distributor:
  kedaAutoscaling:
    kedaRemoteReadURL: http://...

ruler:
  kedaAutoscaling:
    kedaRemoteReadURL: http://...

querier:
  kedaAutoscaling:
    kedaRemoteReadURL: http://...

query_frontend:
  kedaAutoscaling:
    kedaRemoteReadURL: http://...
```


It's less repetitive and slightly more maintainable if we have a single `kedaAutoscaling` section top-level in the values.yaml with all the common parameters:

```yaml
  kedaAutoscaling:
    prometheusAddress: http://... # I didn't like kedaKemoteReadURL because keda is repaeated and the URL isn't a remote-read per-se, it's a prometheus URL, which supports regular reads AND remote-read
    customHeaders:
      {}
    pollingInterval: 10
```

WDYT? I'm also slightly inclined to do this change in a follow-up PR. It will be a breaking change, but that's ok since this is an experimental feature.

Comment (beatkind) [reply operations/helm/charts/mimir-distributed/templates/metamonitoring/_helpers.tpl:67]: I added a comment to #7368 regarding that topic. I agree on adding this option in another PR. I will for now just go a head and not parse anything here.

Comment (beatkind) [reply operations/helm/charts/mimir-distributed/templates/metamonitoring/_helpers.tpl:67]: @dimitarvdimitrov I pushed my current working state. Regarding this parsing, I am still unsure what to do here. Sorry if I misunderstand you:

Using `{{- (.remote).url -}}` would result in `https://mimir.example.com/api/v1/push` which I think is not the correct Read URL.

Comment (dimitarvdimitrov) [reply operations/helm/charts/mimir-distributed/templates/metamonitoring/_helpers.tpl:67]: that's a really good point. The metamonitoring endpoint should end in `/push`, so reusing that directly isn't going to work. 

In this case parsing the URL as you had it probably makes the most sense, but it's rather limiting (wouldn't work with vanila Prometheus because that serves queries under `/api/v1/query`, not `/prometheus/api/v1/query`). I think as a first iteration it's ok to merge as-is and clarify this in the doc string above `kedaAutoscaling`. 

We can iterate on it as part of https://github.com/grafana/mimir/issues/7368

File: operations/helm/charts/mimir-distributed/templates/query-frontend/query-frontend-so.yaml
Comment (dimitarvdimitrov) [inline operations/helm/charts/mimir-distributed/templates/query-frontend/query-frontend-so.yaml:28]: the old query is still here, can you remove it? Same goes for the memory query below

File: operations/helm/charts/mimir-distributed/CHANGELOG.md
Comment (dimitarvdimitrov) [inline operations/helm/charts/mimir-distributed/CHANGELOG.md:31]:
```suggestion
* [FEATURE] Added experimental feature for deploying [KEDA](https://keda.sh) ScaledObjects as part of the helm chart for the components: distributor, querier, query-frontend and ruler. Autoscaling can be enabled via `distributor.kedaAutoscaling`, `ruler.kedaAutoscaling`, `query_frontend.kedaAutoscaling`, and `querier.kedaAutoscaling`. Requires metamonitoring, for more details on metamonitoring see [Monitor the health of your system](https://grafana.com/docs/helm-charts/mimir-distributed/latest/run-production-environment-with-helm/monitor-system-health/). See [grafana/mimir#7367](https://github.com/grafana/mimir/issues/7367) for a migration procedure. #7282
```

the linter is [complaining](https://github.com/grafana/mimir/actions/runs/7885524509/job/21519365699?pr=7282) that we're using base URLs

**Review by dimitarvdimitrov — COMMENTED** [review]
just one comment on the querier SO and a suggestion to satisfy the linter

File: operations/helm/charts/mimir-distributed/templates/metamonitoring/_helpers.tpl
Comment (beatkind) [reply operations/helm/charts/mimir-distributed/templates/metamonitoring/_helpers.tpl:67]: Alright, thanks for the discussion!

File: operations/helm/charts/mimir-distributed/templates/query-frontend/query-frontend-so.yaml
Comment (beatkind) [reply operations/helm/charts/mimir-distributed/templates/query-frontend/query-frontend-so.yaml:28]: Copy and paste error, sorry for that

**Comment by beatkind** [comment]
@dimitarvdimitrov already on my way

**Comment by dimitarvdimitrov** [comment]
I believe this still needs one more `make build-helm-tests`

**Pull request merged**
